In [2]:
words = open('../names.txt', 'r').read().splitlines()

In [3]:
import torch

In [4]:
N = torch.zeros(27, 27, 27, dtype=torch.int32)

In [5]:
chars = sorted(list(set(''.join(['.'] + words))))

itos = {i:s for i, s in enumerate(chars)}
stoi = {i:s for s, i in itos.items()}

In [6]:
for word in words:
  chs = ['.'] * 2 + list(word) + ['.'] * 2
  for ch1, ch2, ch3 in zip(chs, chs[1:], chs[2:]):
    ix1 = stoi[ch1]
    ix2 = stoi[ch2]
    ix3 = stoi[ch3]
    N[ix1, ix2, ix3] += 1

In [7]:
P = (N+1).float()
P /= P.sum(2, keepdim=True)

In [8]:
g = torch.Generator().manual_seed(2147483647)

for i in range(5):
  out = []
  ix1 = 0
  ix2 = 0
  while True:
    p = P[ix1, ix2]
    ix_new = torch.multinomial(p, num_samples=1, replacement=True, generator=g).item()
    out.append(itos[ix_new])
    ix1 = ix2
    ix2 = ix_new
    if ix_new == 0:
      break
  print(''.join(out))


miq.
axx.
mereyannyaar.
knooraen.
el.


In [10]:
log_likelihood = 0.0
n = 0

for word in words:
  chs = ['.'] * 2 + list(word) + ['.'] * 2
  for ch1, ch2, ch3 in zip(chs, chs[1:], chs[2:]):
    ix1 = stoi[ch1]
    ix2 = stoi[ch2]
    ix3 = stoi[ch3]
    prob = P[ix1, ix2, ix3]
    log_likelihood += torch.log(prob)
    n += 1
nll = -log_likelihood
print(f'{nll=:.2f}')
print(f'{nll/n:.2f}')

nll=505260.75
1.94


# Evaluation of trigram counts

average loss: 1.94

Improvement compared to bigram model: 2.45-1.94 = 0.51